### PR TITLE
Ormolu formatting

### PR DESCRIPTION
**Pull request by Munksgaard**
The purpose of this PR is to apply a uniform style to all the Haskell code in Futhark. I've chosen ormolu as the tool to do this, because it is simple to use, is opinionated, and (as of 0.1.2.0) it is idempotent for the code currently in Futhark.

More information about ormolu can be found here: https://github.com/tweag/ormolu. Crucially, this page shows how to set up emacs integration: https://github.com/vyorkin/ormolu.el

For more discussion about code formatting and ormolu, take a look at #42.

### CODE REVIEW

**Comment by Munksgaard** [comment]
Of course, merging this will wreak havoc on all other pull requests and branches branches currently in use, but if we want to do this, we may as well get over that hurdle sooner rather than later.

**Comment by Munksgaard** [comment]
There seems to be an issue with reading non-ascii characters if the locale is not set correctly. As far as I can tell though, the issue is with GHC, not with ormolu as such. More information here: https://github.com/tweag/ormolu/issues/38

**Comment by athas** [comment]
If we do this, then STYLE.md should be replaced with instructions on how to re-run Ormolu.

**Comment by Munksgaard** [comment]
Good point, I'll try to update STYLE.md. However, I don't believe we can completely get rid of STYLE.md, since Ormolu doesn't handle things like rewriting ifs as pattern matches, preferring function pattern matching to `case`, comment best-practices, naming and so on. 

Ormolu also doesn't automatically wrap lines at 80 characters (or any character count, for that matter). It is up the the user to introduce the necessary line breaks, which ormolu will then use as a guide to format the surrounding expressions. For instance, given this long type declaration

```
onKernels :: (SegOp SegLevel KernelsMem -> ReuseAllocsM (SegOp SegLevel KernelsMem)) -> Stms KernelsMem -> ReuseAllocsM (Stms KernelsMem)
```

The user can introduce a single newline

```
onKernels :: (SegOp SegLevel KernelsMem -> ReuseAllocsM (SegOp SegLevel KernelsMem)) -> Stms KernelsMem -> 
  ReuseAllocsM (Stms KernelsMem)
```

and Ormolu will then re-format it like this

```
onKernels ::
  (SegOp SegLevel KernelsMem -> ReuseAllocsM (SegOp SegLevel KernelsMem)) ->
  Stms KernelsMem ->
  ReuseAllocsM (Stms KernelsMem)
```

**Comment by Munksgaard** [comment]
@athas Well, should we?

File: STYLE.md [hunk 1]
```diff
@@ -18,206 +18,150 @@ guidelines on some of the principles underlying the compiler design.
...
-pointless use of alignment, and are retained to serve as a reminder of
-this.
+This command can be used to format all Haskell files in Futhark, while
+checking that the formatting is idempotent:
 
-If in doubt, don't align.  Spurious alignment makes the code feel weird
-and off-key, and it can be remarkably ugly when not maintained.  Alignment
-in expressions is usually a bad idea.
+```
+find src unittests -name '*.hs' -print -execdir ormolu --mode inplace --check-idempotence {} +
```
Comment (athas) [inline STYLE.md:53]: Can you put a script in `tools/` that does this, like we have for style checking?

**Comment by athas** [comment]
According to `git blame`, pretty much every bug in the compiler is now your fault.

**Comment by Munksgaard** [comment]
`git blame -w` is better, though not perfect...

**Pull request merged**